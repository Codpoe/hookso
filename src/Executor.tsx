import React, { useMemo, useEffect } from 'react';
import { Store } from './Store';

export interface ExecutorProps {
  store: Store;
}

export const Executor: React.FC<ExecutorProps> = props => {
  const { store } = props;
  const state = store.hook(store.params);

  useMemo(() => {
    store.setState(state);
  }, [store]); // eslint-disable-line react-hooks/exhaustive-deps

  useEffect(() => {
    store.setState(state);
  }, [store, state]);

  return null;
};
