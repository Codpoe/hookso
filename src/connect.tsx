import React, { useMemo, useCallback, useRef } from 'react';
import { useStore } from './useStore';
import { StoreKey, MapStateToProps, HOC, DepsFn } from './types';

export function connect<P, K = any>(
  key: StoreKey,
  mapStateToProps: MapStateToProps<K, P> = s => s as any
): HOC<P> {
  const wrap: HOC<P> = Component => {
    const Wrapper: React.FC<any> = props => {
      const injectPropsRef = useRef<P>();

      const depsFn = useCallback<DepsFn<K>>(state => {
        injectPropsRef.current = mapStateToProps(state);
        return Object.values(injectPropsRef.current);
      }, []);
      const state = useStore(key, depsFn);

      useMemo(() => {
        injectPropsRef.current = mapStateToProps(state);
        // eslint-disable-next-line react-hooks/exhaustive-deps
      }, []);

      return <Component {...props} {...injectPropsRef.current} />;
    };

    Wrapper.displayName = `${
      Component.displayName || 'AnonymousComponent'
    }Wrapper`;

    return Wrapper;
  };

  return wrap;
}
