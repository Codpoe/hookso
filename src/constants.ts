import { createContext } from 'react';
import { Store } from './Store';
import { StoreKey } from './types';

export const StoresContext = createContext<Record<StoreKey, Store>>({});

export const DEFAULT_STORE_STATE = Symbol('DEFAULT_STORE_STATE') as any;
