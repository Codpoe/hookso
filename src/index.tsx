export * from './Store';
export * from './Executor';
export * from './provide';
export * from './inject';
export * from './use';
export * from './types';
