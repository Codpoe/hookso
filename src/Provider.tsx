import React, { useContext, useMemo, useRef } from 'react';
import { Store } from './Store';
import { Executor } from './Executor';
import { StoresContext } from './constants';
import { FacadeStore, StorKey } from './types';

export interface ProviderProps {
  stores: Pick<FacadeStore<any, any>, 'key' | 'hook' | 'params'>[];
}

export const Provider: React.FC<ProviderProps> = props => {
  const { stores: facadeStores, children } = props;

  const parentStores = useContext(StoresContext);
  const ownStoresRef = useRef<Record<StorKey, Store>>({});

  const stores = useMemo(() => {
    // proto chain
    const _stores: Record<StorKey, Store> = Object.create(parentStores || null);

    return facadeStores.reduce((acc, { key, hook, params }) => {
      if (ownStoresRef.current[key as any]) {
        acc[key as any] = ownStoresRef.current[key as any];
      } else {
        acc[key as any] = new Store(hook, params);
        ownStoresRef.current[key as any] = acc[key as any];
      }
      return acc;
    }, _stores);
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [parentStores, ...facadeStores]);

  return (
    <>
      {facadeStores.map(({ key }, index) => (
        <Executor key={index} store={stores[key as any]} />
      ))}
      <StoresContext.Provider value={stores}>{children}</StoresContext.Provider>
    </>
  );
};
