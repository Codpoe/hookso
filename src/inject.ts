import { useContext, useState, useRef, useEffect } from 'react';
import { StoresContext, DEFAULT_STORE_STATE } from './constants';
import { StorKey, UseStore, Subscriber } from './types';

export function inject<K = any>(key: StorKey): UseStore<K> {
  const useStore: UseStore<K> = depsFn => {
    const store = useContext(StoresContext)[key as any];
    const [state, setState] = useState(() => {
      if (!store) {
        return DEFAULT_STORE_STATE;
      }
      return store.state;
    });
    const depsRef = useRef<any[]>([]);

    useEffect(() => {
      const subscriber: Subscriber = value => {
        if (!depsFn) {
          setState(value);
          return;
        }

        const oldDeps = depsRef.current;
        const newDeps = depsFn(value);

        if (
          oldDeps.length !== newDeps.length ||
          oldDeps.some((dep, index) => dep !== newDeps[index])
        ) {
          setState(value);
        }

        depsRef.current = newDeps;
      };

      return store?.subscribe(subscriber);
    }, [store, depsFn]);

    if (state === DEFAULT_STORE_STATE) {
      throw new Error(
        'No context found. You need to render a Provider of this store.'
      );
    }

    return state;
  };

  return useStore;
}
