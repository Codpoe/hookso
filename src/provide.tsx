import React, { useContext, useMemo } from 'react';
import { Store } from './Store';
import { Executor } from './Executor';
import { StoresContext } from './constants';
import { StorKey, StoreHook } from './types';

export function provide<T, K>(
  key: StorKey,
  hook: StoreHook<undefined, K>
): React.ComponentType<any>;
export function provide<T, K>(
  key: StorKey,
  hook: StoreHook<T, K>,
  params: T
): React.ComponentType<any>;
export function provide<T, K>(
  key: StorKey,
  hook: StoreHook<T | undefined, K>,
  params?: T
): React.ComponentType<any> {
  const Provider: React.FC = props => {
    const { children } = props;
    const parentStores = useContext(StoresContext);
    const store = useMemo(() => new Store(hook, params), []);
    const stores = useMemo(() => {
      // proto chain
      const _stores: Record<StorKey, Store> = Object.create(
        parentStores || null
      );
      // fix: Type 'symbol' cannot be used as an index type
      _stores[key as any] = store;
      return _stores;
    }, [parentStores, store]);

    return (
      <>
        {/* 需要先执行 Executor 获取 hook 的初始 state */}
        <Executor store={store} />
        <StoresContext.Provider value={stores}>
          {children}
        </StoresContext.Provider>
      </>
    );
  };

  return Provider;
}
