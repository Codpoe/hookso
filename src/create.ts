import { provide } from './provide';
import { useStore } from './useStore';
import { connect } from './connect';
import { StoreHook, CreateResult, DepsFn, MapStateToProps } from './types';

export function create<T, K>(hook: StoreHook<undefined, K>): CreateResult<K>;
export function create<T, K>(hook: StoreHook<T, K>, params: T): CreateResult<K>;
export function create<T, K>(
  hook: StoreHook<T | undefined, K>,
  params?: T
): CreateResult<K> {
  const key = Symbol(hook.name || 'AnonymousStore');
  return {
    Provider: provide(key, hook, params),
    // eslint-disable-next-line react-hooks/rules-of-hooks
    use: (depsFn?: DepsFn<K>) => useStore(key, depsFn),
    connect: <P>(mapStateToProps?: MapStateToProps<K, P>) =>
      connect<P, K>(key, mapStateToProps),
  };
}
