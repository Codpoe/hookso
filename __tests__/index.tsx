import React, { useState, useCallback } from 'react';
import { render, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import { create, Provider } from '../src';

const useCounter = (initial?: number) => {
  const [count, setCount] = useState(initial ?? 0);
  const decrement = useCallback(() => setCount(prev => prev - 1), []);
  const increment = useCallback(() => setCount(prev => prev + 1), []);

  return { count, decrement, increment };
};

const counterStore = create(useCounter);
const counterStoreWithParams = create(useCounter, 10);

const App: React.FC = () => {
  const { count, decrement, increment } = counterStore.use();
  return (
    <div>
      <p>{count}</p>
      <button onClick={decrement}>Decrement</button>
      <button onClick={increment}>Increment</button>
    </div>
  );
};

const AppWithDeps: React.FC = () => {
  const { count, decrement, increment } = counterStore.use(() => []);
  return (
    <div>
      <p>{count}</p>
      <button onClick={decrement}>Decrement</button>
      <button onClick={increment}>Increment</button>
    </div>
  );
};

const AppWithParams: React.FC = () => {
  const { count, decrement, increment } = counterStoreWithParams.use();
  return (
    <div>
      <p>{count}</p>
      <button onClick={decrement}>Decrement</button>
      <button onClick={increment}>Increment</button>
    </div>
  );
};

const ConnectedApp = counterStore.connect(state => ({
  n: state.count,
}))((props: { n: number }) => <div>{props.n}</div>);

test('create & use', () => {
  const renderResult = render(
    <counterStore.Provider>
      <App />
    </counterStore.Provider>
  );

  expect(renderResult.asFragment()).toMatchSnapshot();
  expect(renderResult.getByText('0')).toBeTruthy();
  fireEvent.click(renderResult.getByText('Decrement'));
  expect(renderResult.asFragment()).toMatchSnapshot();
  expect(renderResult.queryByText('-1')).toBeTruthy();
});

test('create with params', () => {
  const renderResult = render(
    <counterStoreWithParams.Provider>
      <AppWithParams />
    </counterStoreWithParams.Provider>
  );

  expect(renderResult.asFragment()).toMatchSnapshot();
  expect(renderResult.getByText('10')).toBeTruthy();
  fireEvent.click(renderResult.getByText('Increment'));
  expect(renderResult.asFragment()).toMatchSnapshot();
  expect(renderResult.queryByText('11')).toBeTruthy();
});

test('use with deps', () => {
  const renderResult = render(
    <counterStore.Provider>
      <AppWithDeps />
    </counterStore.Provider>
  );

  expect(renderResult.asFragment()).toMatchSnapshot();
  expect(renderResult.getByText('0')).toBeTruthy();
  fireEvent.click(renderResult.getByText('Decrement'));
  expect(renderResult.asFragment()).toMatchSnapshot();
  expect(renderResult.queryByText('0')).toBeTruthy();
});

test('connect', () => {
  const renderResult = render(
    <counterStore.Provider>
      <ConnectedApp />
    </counterStore.Provider>
  );

  expect(renderResult.asFragment()).toMatchSnapshot();
  expect(renderResult.getByText('0')).toBeTruthy();
});

test('Provider', () => {
  const renderResult = render(
    <Provider stores={[counterStore]}>
      <App />
    </Provider>
  );

  expect(renderResult.asFragment()).toMatchSnapshot();
  expect(renderResult.getByText('0')).toBeTruthy();
  fireEvent.click(renderResult.getByText('Decrement'));
  expect(renderResult.asFragment()).toMatchSnapshot();
  expect(renderResult.queryByText('-1')).toBeTruthy();
});
