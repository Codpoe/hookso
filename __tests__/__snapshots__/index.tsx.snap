// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Provider 1`] = `
<DocumentFragment>
  <div>
    <p>
      0
    </p>
    <button>
      Decrement
    </button>
    <button>
      Increment
    </button>
  </div>
</DocumentFragment>
`;

exports[`Provider 2`] = `
<DocumentFragment>
  <div>
    <p>
      -1
    </p>
    <button>
      Decrement
    </button>
    <button>
      Increment
    </button>
  </div>
</DocumentFragment>
`;

exports[`connect 1`] = `
<DocumentFragment>
  <div>
    0
  </div>
</DocumentFragment>
`;

exports[`create & use 1`] = `
<DocumentFragment>
  <div>
    <p>
      0
    </p>
    <button>
      Decrement
    </button>
    <button>
      Increment
    </button>
  </div>
</DocumentFragment>
`;

exports[`create & use 2`] = `
<DocumentFragment>
  <div>
    <p>
      -1
    </p>
    <button>
      Decrement
    </button>
    <button>
      Increment
    </button>
  </div>
</DocumentFragment>
`;

exports[`create with params 1`] = `
<DocumentFragment>
  <div>
    <p>
      10
    </p>
    <button>
      Decrement
    </button>
    <button>
      Increment
    </button>
  </div>
</DocumentFragment>
`;

exports[`create with params 2`] = `
<DocumentFragment>
  <div>
    <p>
      11
    </p>
    <button>
      Decrement
    </button>
    <button>
      Increment
    </button>
  </div>
</DocumentFragment>
`;

exports[`use with deps 1`] = `
<DocumentFragment>
  <div>
    <p>
      0
    </p>
    <button>
      Decrement
    </button>
    <button>
      Increment
    </button>
  </div>
</DocumentFragment>
`;

exports[`use with deps 2`] = `
<DocumentFragment>
  <div>
    <p>
      0
    </p>
    <button>
      Decrement
    </button>
    <button>
      Increment
    </button>
  </div>
</DocumentFragment>
`;
